File: 2022/puzzle10.py
def run_program(commands):
    command_list = {'noop': (1, lambda x, _: x),
                    'addx': (2, lambda x, y: x + y)}
    x, clock = 1, 1
    program_state = [(0, 0)]

    for command in commands:
        toks = command.split()
        op = toks[0]
        arg = int(toks[1]) if len(toks) == 2 else None
        x = command_list[op][1](x, arg)
        clock += command_list[op][0]
        program_state.append((clock, x))

    return program_state


def state_at(program_state, t):
    for i in range(len(program_state) - 1):
        if program_state[i][0] <= t < program_state[i + 1][0]:
            return program_state[i][1]
    return program_state[-1][1]


def solve_part1(program_state: list):
    part1 = sum([state_at(program_state, clock) *
                clock for clock in [20, 60, 100, 140, 180, 220]])
    print("Part 1: {}".format(part1))


def solve_part2(program_state: list):
    width, height = 40, 6
    sprite_pos = [state_at(program_state, pixel+1)
                  for pixel in range(width * height)]
    screen = ['#' if sprite_pos[pixel]-1 <=
              (pixel % width) <= sprite_pos[pixel]+1 else '.' for pixel in range(width * height)]

    print("Part 2:")
    # print 6 lines of screen at 40 characters each line
    for i in range(height):
        print(''.join(screen[i*width:(i+1)*width]))


with open(r"2022\puzzle10.txt") as f:
    data = f.read().splitlines()

program_state = run_program(data)

solve_part1(program_state)
solve_part2(program_state)

File: 2022/y2022/src/main/java/com/chasingdings/y2022/Puzzle10.java
package com.chasingdings.y2022;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.function.BiFunction;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class Puzzle10 extends AbstractPuzzle {
    private static final String DATA_FILE = "2022\\puzzle10.txt";

    private List<State> programState;

    @Override
    public Object solve1(String content) {
        return IntStream.range(0, 6)
                .map(clock -> clock * 40 + 20)
                .map(clock -> stateAt(programState, clock).x * clock)
                .sum();
    }

    @Override
    public Object solve2(String content) {
        final int width = 40;
        final int height = 6;
        
        final var spritePos = IntStream.range(0, width * height)
                .mapToObj(pixel -> stateAt(programState, pixel + 1))
                .collect(Collectors.toList());
        final var screen = IntStream.range(0, width * height)
                .mapToObj(pixel -> spritePos.get(pixel).x - 1 <= (pixel % width)
                        && (pixel % width) <= spritePos.get(pixel).x + 1 ? "#" : ".")
                .collect(Collectors.joining());

                return "\n" + IntStream.range(0, height)
                .mapToObj(i -> screen.substring(i * width, (i + 1) * width))
                .collect(Collectors.joining("\n"));
    }

    private List<State> runProgram(List<String> commands) {
        List<State> programState = new ArrayList<>();
        programState.add(new State(0, 0));

        var currentState = new State(1, 1);

        for (var command : commands) {
            var toks = command.split(" ");
            var arg = toks.length == 2 ? Integer.parseInt(toks[1]) : null;

            currentState = OP_MAP.get(toks[0]).execute(currentState, arg);
            programState.add(currentState);
        }

        return programState;
    }

    private State stateAt(List<State> programState, int t) {
        for (int i = 0; i < programState.size() - 1; i++) {
            if (programState.get(i).clock <= t && t < programState.get(i + 1).clock) {
                return programState.get(i);
            }
        }

        return programState.get(programState.size() - 1);
    }

    private final Map<String, Instruction> OP_MAP = Map.of(
            "noop", new Instruction(1, (x, y) -> x),
            "addx", new Instruction(2, (x, y) -> x + y));

    class Instruction {
        int clockCycles;
        BiFunction<Integer, Integer, Integer> func;

        public Instruction(int clockCycles, BiFunction<Integer, Integer, Integer> func) {
            this.clockCycles = clockCycles;
            this.func = func;
        }

        public State execute(State state, Integer arg) {
            return new State(state.clock + clockCycles, func.apply(state.x, arg));
        }
    }

    class State {
        int clock;
        int x;

        public State(int clock, int x) {
            this.clock = clock;
            this.x = x;
        }
    }

    @Override
    public void preprocess(String content) {
        programState = runProgram(getInputDataByLine(content));
    }

    @Override
    public String getDataFilePath() {
        return DATA_FILE;
    }

    @Override
    public String getPuzzleName() {
        return "Day 10 - Cathode-Ray Tube";
    }
}
